src/Features/Core/Portable/SignatureHelp/SymbolSignatureHelpProvider.cs

```diff
@@ -23,52 +23,66 @@ internal abstract partial class AbstractSignatureHelpProvider : ISignatureHelpPr
-            IList<SignatureHelpItem> items, TextSpan applicableSpan, SignatureHelpState state)
+        public override async Task ProvideSignaturesAsync(SignatureContext context)
         {
-            if (items == null || !items.Any() || state == null)
+            await ProvideSignaturesWorkerAsync(context).ConfigureAwait(false);
+            if (context.Items.Count == 0)
             {
-                return null;
+                return;
             }
 
-            items = Filter(items, state.ArgumentNames);
-            return new SignatureHelpItems(items, applicableSpan, state.ArgumentIndex, state.ArgumentCount, state.ArgumentName);
-        }
-
-        private static IList<SignatureHelpItem> Filter(IList<SignatureHelpItem> items, IEnumerable<string> parameterNames)
-        {
-            if (parameterNames == null)
+            var relatedDocuments = context.Document.GetLinkedDocumentIds();
+            if (!relatedDocuments.Any())
             {
-                return items.ToList();
+                return;
             }
 
-            var filteredList = items.Where(i => Include(i, parameterNames)).ToList();
-            return filteredList.Count == 0 ? items.ToList() : filteredList;
-        }
+            var document = context.Document;
+            var position = context.Position;
+            var trigger = context.Trigger;
+            var cancellationToken = context.CancellationToken;
 
-        private static bool Include(SignatureHelpItem item, IEnumerable<string> parameterNames)
-        {
-            var itemParameterNames = item.Parameters.Select(p => p.Name).ToSet();
-            return parameterNames.All(itemParameterNames.Contains);
-        }
+            var relatedDocumentsAndItems = await GetItemsForRelatedDocuments(context, relatedDocuments).ConfigureAwait(false);
+            var candidateLinkedProjectsAndSymbolSets = await ExtractSymbolsFromRelatedItems(position, relatedDocumentsAndItems, cancellationToken).ConfigureAwait(false);
 
-        public async Task<SignatureHelpState> GetCurrentArgumentStateAsync(Document document, int position, TextSpan currentSpan, CancellationToken cancellationToken)
-        {
-            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            return GetCurrentArgumentState(root, position, document.GetLanguageService<ISyntaxFactsService>(), currentSpan, cancellationToken);
+            var totalProjects = candidateLinkedProjectsAndSymbolSets.Select(c => c.Item1).Concat(document.Project.Id);
+
+            var semanticModel = await document.GetSemanticModelForSpanAsync(new TextSpan(position, 0), cancellationToken).ConfigureAwait(false);
+            var compilation = semanticModel.Compilation;
+            var finalItems = new List<SignatureHelpItem>();
+            foreach (var item in context.Items)
+            {
+                var symbolKey = ((SymbolKeySignatureHelpItem)item).SymbolKey;
```

Comment:
Agreed. This is not new code though, it was relocated. I'm going to revisit all of this before I check in.